sklearn/decomposition/dict_learning.py

```diff
@@ -343,35 +352,32 @@ def _update_dict(dictionary, Y, code, verbose=False, return_r2=False,
         Updated dictionary.
 
     """
-    n_components = len(code)
-    n_samples = Y.shape[0]
+    n_features, n_components = B.shape
     random_state = check_random_state(random_state)
     # Residuals, computed 'in-place' for efficiency
-    R = -np.dot(dictionary, code)
-    R += Y
+    R = -np.dot(dictionary, A)
+    R += B
     R = np.asfortranarray(R)
-    ger, = linalg.get_blas_funcs(('ger',), (dictionary, code))
+    ger, = linalg.get_blas_funcs(('ger',), (dictionary, A))
     for k in range(n_components):
         # R <- 1.0 * U_k * V_k^T + R
-        R = ger(1.0, dictionary[:, k], code[k, :], a=R, overwrite_a=True)
-        dictionary[:, k] = np.dot(R, code[k, :].T)
+        R = ger(1.0, dictionary[:, k], A[k, :], a=R, overwrite_a=True)
+        dictionary[:, k] = np.dot(R, A[k, :])
```

Comment:
I tried working through this on paper and I think the batch update ends up different by a factor of `dot(code.T, code)`.

The old implementation seems incorrect to me; your implementation corresponds to the online update in the paper, applied over the entire dataset.

Previously we had:

```
R = X - DCt   # dimensions here are transposed compared to sklearn conventions
D[:, k] = R * C[k] = (X - DCt) C[k]
```
Now, by forming A and B, we have
```
R = B - DA = XC - DCtC = (X - DCt)C
D[:, k] = R Ak = R CtC[k] = (X - DCt) CCt C[k]
```

Am I missing something or was this indeed a bug previously?